Add doc and example app for config search path (#1468)
facebookresearch/hydra: website/docs/advanced/search_path.md

title: Config Search Path
---

import {ExampleGithubLink} from "@site/src/components/GithubLink"

The Config Search Path is a list of paths that Hydra searches in order to find configs. It is similar to
the Python `PYTHONPATH`. 
 - When a config is requested, The first matching config in the search path is used.
 - Each search path element has a schema prefix such as `file://` or `pkg://` that is corresponding to a `ConfigSourcePlugin`.

You can inspect the search path and the configurations loaded by Hydra by turning on verbose logging for the `hydra` logger:

```bash

> $ python my_app.py hydra.verbose=hydra

Reviewer: I think we should already suggest `--info config` instead of hydra.verbose=hydra.
When we have --info searchpath we can change to that.

```

There are a few ways to modify the config search path, enabling Hydra to access configuration in 
different locations.
Use a combination of the methods described below:

#### Using `@hydra.main()`
Using the  `config_path` parameter `@hydra.main()`.  The `config_path` is relative to location of the Python script.

#### Overriding `hydra.searchpath` config

<ExampleGithubLink text="Example application" to="examples/advanced/config_search_path"/>

In some cases you may want to add multiple locations to the search path. 
For example, an app may want to read the configs from an additional Python module or 
an additional directory on the file system.  
Configure this using `hydra.searchpath` in your primary config or your command line.
:::info
hydra.searchpath can **only** be configured in the primary config. Attempting  to configure it in other configs will result in an error.
:::

> Let's say we have the following configuration structure and application code:

Reviewer: ```suggestion
In this example, we add a second config directory - `additional_conf`, next to the first config directory:
```

> Let's say we have the following configuration structure and application code:
> <div className="row">
> <div className="col col--4">
> 
> ```bash
> ├── __init__.py
> ├── additonal_conf
> │   ├── __init__.py
> │   └── dataset
> │       └── imagenet.yaml
> ├── conf
> │   ├── config.yaml
> │   └── dataset
> │       └── cifar10.yaml
> └── my_app.py
> ```
> </div>
> <div className="col  col--8">

Reviewer: Swap the order of the config directories:
```suggestion
Let's say we have the following configuration structure and application code:
<div className="row">
<div className="col col--4">

```bash
├── __init__.py
├── conf
│   ├── config.yaml
│   └── dataset
│       └── cifar10.yaml
├── additonal_conf
│   ├── __init__.py
│   └── dataset
│       └── imagenet.yaml
└── my_app.py
```
</div>
<div className="col  col--8">
```

```python title="my_app.py"

@hydra.main(config_path="conf", config_name="config")
def my_app(cfg: DictConfig) -> None:
    print(OmegaConf.to_yaml(cfg))

if __name__ == "__main__":
    my_app()
```
</div>
</div>

`conf/config.yaml` is the primary config for `my_app.py`. To reference `additonal_conf/dataset/imagenet` 
in `config.yaml`, we can add `additonal_conf` to 
`hydra.searchpath` so Hydra could discover the configurations under `additonal_conf`.

<div className="row">
<div className="col col--7">

```yaml title="config.yaml"
defaults:

>   - dataset: imagenet

Reviewer: make the default cifar10, and then make a point of how you can override the dataset to imagenet, which is provided by additional_conf.

hydra:
  searchpath:
    - pkg://additonal_conf
    # You can also use file based schema:
    # - file:///etc/my_app
    # - file://${env:HOME}/.my_app
```

</div>

<div className="col  col--5">

```python title="my_app.py output"
dataset:
  name: imagenet
  path: /datasets/imagenet

```
</div>
</div>

> In addition to override `hydra.searchpath` in primary config, one can also override from the command line.
> 

Reviewer: ```suggestion
`hydra.searchpath` can be defined or overridden via the command line as well:
```

```bash title="command line override"
python my_app.py 'hydra.searchpath=[pkg://additonal_conf]'
```

#### Overriding `--config-dir` from the command line
This is a less flexible alternative to `hydra.searchpath`. 

> See this [page](/docs/advanced/hydra-command-line-flags) for more info.

Reviewer: Maybe we should deprecate it as as hydra.searchpath looks like it can completely replace it.

Author: Removed this from the doc for now :)

title: Config Search Path
---

import {ExampleGithubLink} from "@site/src/components/GithubLink"

The Config Search Path is a list of paths that Hydra searches in order to find configs. It is similar to
the Python `PYTHONPATH`. 
 - When a config is requested, The first matching config in the search path is used.

>  - Each search path element has a schema prefix such as `file://` or `pkg://` that is corresponding to a `ConfigSourcePlugin`.

Reviewer: I think we need to explain each of the prefixes.

- file:// is a file system path. it should normally be absolute, if its relative I _think_ it's relative to cwd. I think separator is forward slash (even on Windows).
- pkg:// is a python module. it should be importable and should contain and `__init__.py`. separator is a `.`.

Double check those statements and update the docs with the info.

title: Config Search Path
---

import {ExampleGithubLink} from "@site/src/components/GithubLink"

The Config Search Path is a list of paths that Hydra searches in order to find configs. It is similar to
the Python `PYTHONPATH`. 
 - When a config is requested, The first matching config in the search path is used.
 - Each search path element has a schema prefix such as `file://` or `pkg://` that is corresponding to a `ConfigSourcePlugin`.

You can inspect the search path and the configurations loaded by Hydra by turning on verbose logging for the `hydra` logger:

```bash

> $ python my_app.py hydra.verbose=hydra

Reviewer: 1. --info config is better.
2. you can add --info searchpath (different pr).
3. the searchpath is incomplete and does not contain the effective searchpath from the config file. You can fix it (another PR).

Reviewer: Additionally, I think we should output a warning of a config searchpath is not pointing to a valid location.
You can fix it in a different PR.

Author: Created an issue for the follow up needed.

```

> There's a few options to modify the config search path so Hydra could discover 
> the configurations. You can pick and choose some or all of the following methods below
> to provide configuration to your apps.

Reviewer: ```suggestion
There's a few options to modify the config search path so Hydra could discover 
the configurations. Use a combination of the methods described below:
```

> ### Set `config_name`/`config_path` in`hydra.main()`

Reviewer: ```suggestion
### Using @hydra.main()
Specify a `config_path` in `@hydra.main()`. This path is relative to ...
```

This is the most common use case: all config files are grouped under one directory and
has a relative path to the application. You can add them to the search path by setting `config_path` 
or `config_name`.

> 

Reviewer: 1. config_name should **absolutely** not be used to add a directory to the search path.
2. style wise, the title looks huge (look at the resulting page).

Author: Fixed the text & downsized the title

### Override `hydra.searchpath` config

<ExampleGithubLink text="Example application" to="examples/advanced/config_search_path"/>

> Sometimes your configurations are not grouped nicely in one folder - 
> you may need to read configurations from multiple locations, so passing one directory to `config_path` 
> won't work. In this case, you can pass in a list of config paths via `hydra.searchpath` in 
> the primary config or override from the command line.

Reviewer: ```suggestion
In some cases you may want to add multiple locations to the search path. For example, an app may want to read the configs from an additional Python module or an additional directory on the file system.  
Configure this using `hydra.searchpath` in your primary config or your command line.

:::info
hydra.searchpath can **only** be configured in the primary config. Attempting  to configure it in other configs will result in an error.
:::
```

Let's say we have the following configuration and application structure:
```bash
$ tree  config_search_path
config_search_path
├── __init__.py
├── conf
│   └── config.yaml
├── dataset
│   ├── cifar10.yaml
│   └── imagenet.yaml
└── my_app.py
```
where `conf/config.yaml` is the primary config for `my_app.py`. However at the same time, we also 
want to reference configuration from `dataset` in `config.yaml`. In this case, we can add `dataset` to 

> `hydra.searchpath` so Hydra could discover the configurations under `dataset`.

Reviewer: You have added a config searthpath inside another config searcpath. this is very bad example for people to follow :).

```yaml title="config.yaml"
defaults:
  - dataset: imagenet

hydra:
  searchpath:
    - pkg://examples.advanced.config_search_path
```

```python title="my_app.py"

@hydra.main(config_path="conf", config_name="config")
def my_app(cfg: DictConfig) -> None:
    print(OmegaConf.to_yaml(cfg))

if __name__ == "__main__":
    my_app()
```

```python title="my_app.py output"
dataset:
  name: imagenet
  path: /datasets/imagenet
```

In addition to override `hydra.searchpath` in primary config, one can also override from the command line.

```bash title="command line override"
python my_app.py hydra.searchpath=[pkg://examples.advanced.config_search_path]
```

### Override `--config-dir` from the command line
This is a less flexible alternative to `hydra.searchpath`. See this [page](/docs/advanced/hydra-command-line-flags) for more info.

### Create `SearchPathPlugin`

<ExampleGithubLink text="ExampleSearchPathPlugin" to="examples/plugins/example_searchpath_plugin/"/>

> Hydra auto discovers all the Plugins that's created under a `hydra_plugins` package. So if modifying `hydra.searchpath` 

Reviewer: First sentence is true but not relevant.

> is not an option for you, you can still create a `SearchPathPlugin` for the configs to be discovered. 

Reviewer: Why would someone want to use SearchPathPlugin over hydra.searchpath?

Author: Added some more context here

title: Config Search Path
---

import {ExampleGithubLink} from "@site/src/components/GithubLink"

The Config Search Path is a list of paths that Hydra searches in order to find configs. It is similar to
the Python `PYTHONPATH`. 
 - When a config is requested, The first matching config in the search path is used.
 - Each search path element has a schema prefix such as `file://` or `pkg://` that is corresponding to a `ConfigSourcePlugin`.

You can inspect the search path and the configurations loaded by Hydra by turning on verbose logging for the `hydra` logger:

```bash
$ python my_app.py hydra.verbose=hydra
```

There's a few options to modify the config search path so Hydra could discover 
the configurations. Use a combination of the methods described below:

#### Using `@hydra.main()`

> Specify a `config_path` in `@hydra.main()`. This path is relative to the application.
> All the configuration under the path will be discovered by Hydra.

Reviewer: ```suggestion
Using the  `config_path` parameter `@hydra.main()`.  The `config_path` is relative to location of the Python script.
```

:::

Let's say we have the following configuration and application structure:
```bash
├── __init__.py
├── additonal_conf
│   ├── __init__.py
│   └── dataset
│       ├── cifar10.yaml
│       └── imagenet.yaml
├── conf
│   └── config.yaml
└── my_app.py
```
where `conf/config.yaml` is the primary config for `my_app.py`. At the same time, we also 
want to reference configuration from `additonal_conf/dataset` in `config.yaml`. In this case, we can add `additonal_conf` to 
`hydra.searchpath` so Hydra could discover the configurations under `additonal_conf`.

```yaml title="config.yaml"
defaults:
  - dataset: imagenet

hydra:
  searchpath:
    - pkg://additonal_conf
```

```python title="my_app.py"

@hydra.main(config_path="conf", config_name="config")
def my_app(cfg: DictConfig) -> None:
    print(OmegaConf.to_yaml(cfg))

if __name__ == "__main__":
    my_app()
```

```python title="my_app.py output"
dataset:
  name: imagenet
  path: /datasets/imagenet
```

In addition to override `hydra.searchpath` in primary config, one can also override from the command line.

```bash title="command line override"

> python my_app.py hydra.searchpath=[pkg://additonal_conf]

Reviewer: While this works, the shell is actually interpreting square brackets and in some rare cases it may cause issues.
```suggestion
python my_app.py 'hydra.searchpath=[pkg://additonal_conf]'
```

```

#### Overriding `--config-dir` from the command line
This is a less flexible alternative to `hydra.searchpath`. 
See this [page](/docs/advanced/hydra-command-line-flags) for more info.

#### Creating a `SearchPathPlugin`

<ExampleGithubLink text="ExampleSearchPathPlugin" to="examples/plugins/example_searchpath_plugin/"/>

> If you do not have control of the primary config file but still want to provide 
> configurations to the end users, you can choose to create a `SearchPathPlugin`.
> Hydra auto discovers all the Plugins that's created under a `hydra_plugins` package. 
> A `SearchPathPlugin` serves the purpose of adding to Hydra's config search path. 
> For more details, please check the example plugin linked above.

Reviewer: ```suggestion
Framework authors may want to add their configs to the search path automatically once their package is installed,
eliminating the need for any actions from the users.
This can be achieved using a `SearchPathPlugin`. Check the example plugin linked above for more details.
```

</div>
</div>

Overriding `dataset=imagenet` from the commandline:

<div className="row">
<div className="col col--6">

```bash title="command line override"
python my_app.py dataset=imagenet

```

</div>

<div className="col  col--6">

```python title="my_app.py output"
dataset:
  name: imagenet
  path: /datasets/imagenet
```
</div>
</div>

In addition to override `hydra.searchpath` in primary config, one can also override from the command line.
`hydra.searchpath` can be defined or overridden via the command line as well:

```bash title="command line override"
python my_app.py 'hydra.searchpath=[pkg://additonal_conf]'
```

#### Overriding `--config-dir` from the command line
This is a less flexible alternative to `hydra.searchpath`. 

> See this [page](/docs/advanced/hydra-command-line-flags) for more info.

Reviewer: Don't remove it yet.
As long as we didn't deprecate it it should be documented here.
